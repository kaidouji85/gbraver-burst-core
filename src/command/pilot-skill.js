// @flow

/**
 * パイロットスキルコマンド
 */
export type PilotSkillCommand = {
  type: 'PilotSkillCommand'
}