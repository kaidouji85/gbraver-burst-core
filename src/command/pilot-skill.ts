/**
 * パイロットスキルコマンド
 */
export type PilotSkillCommand = {
  type: "PILOT_SKILL_COMMAND";
};