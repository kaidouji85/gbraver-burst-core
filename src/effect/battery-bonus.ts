/**
 * バッテリーによるダメージボーナス
 * @param attackBattery 攻撃側が出したバッテリー
 * @param defenseBattery 防御側が出したバッテリー
 * @returns ダメージボーナス
 */
export function batteryBonus(
  attackBattery: number,
  defenseBattery: number,
): number {
  return 100 * Math.max(attackBattery - defenseBattery - 1, 0);
}
