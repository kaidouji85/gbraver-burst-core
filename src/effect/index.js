// @flow

import type {InputCommand} from "./input-command/input-command";
import type {Battle} from "./battle/battle";
import type {TurnChange} from "./turn-change/turn-change";
import type {StartGame} from "./start-game/start-game";
import type {BurstEffect} from "./burst/burst-effect";
import type {GameEnd} from "./game-end/game-end";
import type {BatteryDeclaration} from "./battery-declaration/battery-declaration";
import type {Reflect} from "./reflect/reflect";
import type {UpdateRemainingTurn} from "./update-remaning-turn/update-remaining-turn";

/** ゲーム中に発生する効果をまとめたもの */
export type Effect =
  StartGame |
  GameEnd |
  InputCommand |
  Reflect |
  BatteryDeclaration |
  Battle |
  TurnChange |
  BurstEffect |
  UpdateRemainingTurn;
