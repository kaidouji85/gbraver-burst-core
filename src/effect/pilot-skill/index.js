// @flow

import type {GameState, GameStateX, PilotSkill, PilotSkillEffect, PilotSkillEffectX, PlayerId} from "../..";
import type {BuffPowerSkill, RecoverBatterySkill} from "../../player/pilot";
import {recoverBattery} from "./recover-battery";
import {buffPower} from "./buff-power";

/**
 * パイロットスキルを発動する
 *
 * @param lastState 最新の状態
 * @param invokerId パイロットスキルを発動するプレイヤー
 * @return 更新結果、実行不可能な場合はnullを返す
 */
export function pilotSkill(lastState: GameState, invokerId: PlayerId): ?GameStateX<PilotSkillEffect> {
  const donePilotSkill = pilotSkillEffect(lastState, invokerId);
  if (!donePilotSkill) {
    return null;
  }
  return disablePilotSkill(donePilotSkill);
}

/**
 * パイロットスキルを適用する
 *
 * @param lastState 最新の状態
 * @param invokerId パイロットスキルを発動するプレイヤー
 * @return 更新結果、実行不可能な場合はnullを返す
 */
function pilotSkillEffect(lastState: GameState, invokerId: PlayerId): ?GameStateX<PilotSkillEffect> {
  const invoker = lastState.players.find(v => v.playerId === invokerId);
  if (!invoker) {
    return null;
  }

  if (invoker.pilot.skill.type === 'RecoverBatterySkill') {
    const castedSkill: RecoverBatterySkill = invoker.pilot.skill;
    const updated =  recoverBattery(lastState, invokerId, castedSkill);
    return updated ? upcastPilotSkillEffect(updated) : null;
  }

  if (invoker.pilot.skill.type === 'BuffPowerSkill') {
    const castedSkill: BuffPowerSkill = invoker.pilot.skill;
    const updated = buffPower(lastState, invokerId, castedSkill);
    return updated ? upcastPilotSkillEffect(updated) : null;
  }

  return null;
}

/**
 * パイロットスキルを使用済みにする
 *
 * @param lastState 最新状態
 * @return 更新結果、実行不可能な場合はnullを返す
 */
function disablePilotSkill(lastState: GameStateX<PilotSkillEffect>): ?GameStateX<PilotSkillEffect> {
  const invoker = lastState.players.find(v => v.playerId === lastState.effect.invokerId);
  if (!invoker) {
    return null;
  }

  const updatedInvoker = {
    ...invoker,
    pilot: {
      ...invoker.pilot,
      enableSkill: false
    }
  };
  const updatedPlayers = lastState.players
    .map(v => v.playerId === updatedInvoker.playerId ? updatedInvoker : v);
  return {
    ...lastState,
    players: updatedPlayers
  };
}

/**
 * パイロットスキル発動ステートにアップキャストする
 *
 * @param origin キャスト前
 * @return キャスト結果
 */
function upcastPilotSkillEffect<X: PilotSkill>(origin: GameStateX<PilotSkillEffectX<X>>): GameStateX<PilotSkillEffect> {
  return ((origin: any): GameStateX<PilotSkillEffectX<PilotSkill | typeof origin.effect.skill>>);
}
