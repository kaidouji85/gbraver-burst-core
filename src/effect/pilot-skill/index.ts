import type { PlayerId } from "../../player/player";
import type { GameState, GameStateX } from "../../state/game-state";
import { PlayerState } from "../../state/player-state";
import { batteryBoost } from "./battery-boost";
import { batteryEnchantment } from "./battery-enchantment";
import { buffPower } from "./buff-power";
import { damageHalvedSkill } from "./damage-halved-skill";
import type { PilotSkillEffect } from "./pilot-skill-effect";
import { recoverBattery } from "./recover-battery";

/**
 * パイロットスキルを適用する
 * @param lastState 最新の状態
 * @param invokerId パイロットスキルを発動するプレイヤー
 * @return 更新結果、実行不可能な場合は例外を返す
 */
function pilotSkillEffect(
  lastState: GameState,
  invokerId: PlayerId,
): GameStateX<PilotSkillEffect> {
  const invoker = lastState.players.find((v) => v.playerId === invokerId);
  if (!invoker) {
    throw new Error("not found pilot skill invoker");
  }

  if (invoker.pilot.skill.type === "RecoverBatterySkill") {
    return recoverBattery(lastState, invokerId, invoker.pilot.skill);
  }

  if (invoker.pilot.skill.type === "BuffPowerSkill") {
    return buffPower(lastState, invokerId, invoker.pilot.skill);
  }

  if (invoker.pilot.skill.type === "BatteryEnchantmentSkill") {
    return batteryEnchantment(lastState, invokerId, invoker.pilot.skill);
  }

  if (invoker.pilot.skill.type === "DamageHalvedSkill") {
    return damageHalvedSkill(lastState, invokerId, invoker.pilot.skill);
  }

  if (invoker.pilot.skill.type === "BatteryBoostSkill") {
    return batteryBoost(lastState, invokerId, invoker.pilot.skill);
  }

  throw new Error("not found pilot skill");
}

/**
 * パイロットスキルを使用済みにする
 * @param lastState 最新状態
 * @return 更新結果
 */
function disablePilotSkill(
  lastState: GameStateX<PilotSkillEffect>,
): GameStateX<PilotSkillEffect> {
  const updateInvoker = (invoker: PlayerState) => ({
    ...invoker,
    pilot: { ...invoker.pilot, enableSkill: false },
  });
  const updatedPlayers = lastState.players.map((v) =>
    v.playerId === lastState.effect.invokerId ? updateInvoker(v) : v,
  );
  return { ...lastState, players: updatedPlayers };
}

/**
 * パイロットスキルを発動する
 * @param lastState 最新の状態
 * @param invokerId パイロットスキルを発動するプレイヤー
 * @return 更新結果、実行不可能な場合は例外を返す
 */
export function pilotSkill(
  lastState: GameState,
  invokerId: PlayerId,
): GameStateX<PilotSkillEffect> {
  const donePilotSkill = pilotSkillEffect(lastState, invokerId);
  return disablePilotSkill(donePilotSkill);
}
