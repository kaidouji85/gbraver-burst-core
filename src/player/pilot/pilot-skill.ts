import { z } from "zod";

import {
  BatteryBoostSkill,
  BatteryBoostSkillSchema,
} from "./battery-boost-skill";
import {
  BatteryEnchantmentSkill,
  BatteryEnchantmentSkillSchema,
} from "./battery-enchantment-skill";
import { BuffPowerSkill, BuffPowerSkillSchema } from "./buff-power-skill";
import {
  DamageHalvedSkill,
  DamageHalvedSkillSchema,
} from "./damage-halved-skill";
import {
  RecoverBatterySkill,
  RecoverBatterySkillSchema,
} from "./recover-battery-skill";

/** パイロットスキル */
export type PilotSkill =
  | RecoverBatterySkill
  | BuffPowerSkill
  | BatteryEnchantmentSkill
  | DamageHalvedSkill
  | BatteryBoostSkill;

/** PilotSkill zodスキーマ */
export const PilotSkillSchema = z.union([
  RecoverBatterySkillSchema,
  BuffPowerSkillSchema,
  BatteryEnchantmentSkillSchema,
  DamageHalvedSkillSchema,
  BatteryBoostSkillSchema,
]);
