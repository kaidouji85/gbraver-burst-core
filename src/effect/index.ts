import { z } from "zod";

import { BatteryDeclaration, BatteryDeclarationSchema } from "./battery-declaration/battery-declaration";
import { Battle, BattleSchema } from "./battle/battle";
import { BurstEffect, BurstEffectSchema } from "./burst/burst-effect";
import { GameEnd, GameEndSchema } from "./game-end/game-end";
import { InputCommand, InputCommandSchema } from "./input-command/input-command";
import { PilotSkillEffect, PilotSkillEffectSchema } from "./pilot-skill/pilot-skill-effect";
import { Reflect, ReflectSchema } from "./reflect/reflect";
import { RightItself, RightItselfSchema } from "./right-itself/right-itself";
import { StartGame, StartGameSchema } from "./start-game/start-game";
import { TurnChange, TurnChangeSchema } from "./turn-change/turn-change";
import { UpdateRemainingTurn, UpdateRemainingTurnSchema } from "./update-remaning-turn/update-remaining-turn";

/** ゲーム中に発生する効果をまとめたもの */
export type Effect =
  | StartGame
  | GameEnd
  | InputCommand
  | Reflect
  | BatteryDeclaration
  | Battle
  | TurnChange
  | BurstEffect
  | UpdateRemainingTurn
  | RightItself
  | PilotSkillEffect;

/** Effect zodスキーマ */
export const EffectSchema = z.union([
  StartGameSchema,
  GameEndSchema,
  InputCommandSchema,
  ReflectSchema,
  BatteryDeclarationSchema,
  BattleSchema,
  TurnChangeSchema,
  BurstEffectSchema,
  UpdateRemainingTurnSchema,
  RightItselfSchema,
  PilotSkillEffectSchema,
]);