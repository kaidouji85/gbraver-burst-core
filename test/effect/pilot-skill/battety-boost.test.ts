import {
  ArmdozerEffect,
  EMPTY_ARMDOZER_STATE,
  EMPTY_GAME_STATE,
  EMPTY_PILOT_STATE,
  EMPTY_PLAYER_STATE,
  GameState,
  PlayerState,
} from "../../../src";
import { pilotSkill } from "../../../src/effect/pilot-skill";

/** BatteryRecoverSkip以外の効果 */
const effectOtherThanBatteryRecoverSkip: ArmdozerEffect = {
  type: "CorrectPower",
  power: 1000,
  period: {
    type: "TurnLimit",
    remainingTurn: 1,
  },
};

/** 効果発動プレイヤー */
const invoker: PlayerState = {
  ...EMPTY_PLAYER_STATE,
  playerId: "invoker",
  armdozer: {
    ...EMPTY_ARMDOZER_STATE,
    battery: 0,
    maxBattery: 8,
    effects: [
      effectOtherThanBatteryRecoverSkip,
      effectOtherThanBatteryRecoverSkip,
    ],
  },
  pilot: {
    ...EMPTY_PILOT_STATE,
    skill: {
      type: "BatteryBoostSkill",
      recoverBattery: 5,
    },
    enableSkill: true,
  },
};

/** それ以外のプレイヤー */
const other = {
  ...EMPTY_PLAYER_STATE,
  playerId: "other",
};

/** ゲームステート */
const lastState: GameState = {
  ...EMPTY_GAME_STATE,
  activePlayerId: other.playerId,
  players: [invoker, other],
};

test("バッテリーブーストスキルを正しく適用できる", () => {
  const result = pilotSkill(lastState, invoker.playerId);
  expect(result).toMatchSnapshot();
});
