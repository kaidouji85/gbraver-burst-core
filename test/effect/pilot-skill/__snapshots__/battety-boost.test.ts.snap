// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`バッテリーブーストスキルを正しく適用できる 1`] = `
{
  "activePlayerId": "other",
  "effect": {
    "invokerId": "invoker",
    "name": "PilotSkillEffect",
    "skill": {
      "recoverBattery": 5,
      "type": "BatteryBoostSkill",
    },
  },
  "players": [
    {
      "armdozer": {
        "battery": 5,
        "batteryAutoRecovery": 3,
        "burst": {
          "recoverBattery": 5,
          "turnStartBatteryCorrect": 1,
          "type": "RecoverBattery",
        },
        "effects": [
          {
            "period": {
              "remainingTurn": 1,
              "type": "TurnLimit",
            },
            "power": 1000,
            "type": "CorrectPower",
          },
          {
            "period": {
              "remainingTurn": 1,
              "type": "TurnLimit",
            },
            "power": 1000,
            "type": "CorrectPower",
          },
          {
            "period": {
              "type": "SpecialPeriod",
            },
            "type": "BatteryRecoverSkip",
          },
        ],
        "enableBurst": true,
        "hp": 3000,
        "id": "id",
        "maxBattery": 8,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": false,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "recoverBattery": 5,
          "type": "BatteryBoostSkill",
        },
      },
      "playerId": "invoker",
    },
    {
      "armdozer": {
        "battery": 5,
        "batteryAutoRecovery": 3,
        "burst": {
          "recoverBattery": 5,
          "turnStartBatteryCorrect": 1,
          "type": "RecoverBattery",
        },
        "effects": [],
        "enableBurst": true,
        "hp": 3000,
        "id": "id",
        "maxBattery": 5,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": true,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "recoverBattery": 2,
          "type": "RecoverBatterySkill",
        },
      },
      "playerId": "other",
    },
  ],
}
`;
